cclunit-framework-source/doc/CCLUTMOCKING.md

```diff
@@ -0,0 +1,268 @@
+  
+@param originalName  
+&nbsp;&nbsp;&nbsp;&nbsp;The object to be mocked.  
+@param replaceName  
+&nbsp;&nbsp;&nbsp;&nbsp;The mocked object.  
+  
+Example:  
+call cclutAddMockImplementation("uar_get_code_by", "mock_uar_get_code_by")
+
+**cclutRemoveMockImplementation(originalName = vc)**
+
+Removes a mock implementation.  
+  
+@param originalName  
+&nbsp;&nbsp;&nbsp;&nbsp;The object that is mocked.  
+  
+Example:  
+call cclutRemoveMockImplementation("uar_get_code_by")
+
+**cclutRemoveAllMockImplementations(null)**
+
+Removes all mock implementations.  
+  
+Example:  
+call cclutRemoveAllMockImplementations(null)
+
+**cclutExecuteProgramWithMocks(programName = vc, params = vc, namespace = vc)**
+
+Executes a program with all mocks currently added through cclutAddMockImplementation() and cclutCreateMockTable(). programName is required.  params is a string parameter to be sent directly to the program, so all commas and string delimiters must be specified.  If namespace is omitted, it will default to the PUBLIC namespace.  
+  
+Example:  
+call cclutExecuteProgramWithMocks("ccl_my_program", "\^MINE^, 1.0, ^string parameter^", "MYNAMESPACE")
+
+**cclutRemoveAllMocks**
+
+Removes all mock implementations and mock tables that have been added through the cclutAddMockImplementation() and cclutCreateMockTable() APIs.  This should be called at the completion of a test suite to clean up all mocks.  
+  
+Example:  
+call cclutRemoveAllMocks(null)
+
+## Implementation Notes
+1. cclutRemoveAllMocks should be called as part of the teardown for all tests.  The framework will attempt to clean up any outstanding mocks, but it is good practice to explicitly remove any mocks to ensure that no mocked tables remain in the Oracle instance.
+
+2. The mocked items created through cclutCreateMockTable() and cclutAddMockImplementation will not be applied to children script called from the script-under-test.  Some alternatives would be to mock the child script to return the appropriate data or to mock the child script to execute the real script applying the mocked tables and implementations.
+
+3. The mocked items created through cclutCreateMockTable() and cclutAddMockImplementation will not be applied to statements executed through "call parser()" commands.  An alternative would be to mock the parser() call to validate the correct information is supplied, then perform the appropriate mock versions of the actions the statement would normally perform.
+
+4. The mocking API calls cannot be used from reportwriter sections.  Alternatives would be dependent on the use-case, but take, for example, a dummyt used with a record structure and a call to cclutAddMockData within the detail section in order to add data based on a record structure.  Instead, a FOR loop construct could be leveraged outside the context of a reportwriter section to iterate over the record structure.
```

Comment:
So a couple things.

First off, your example wasn't what I intended, so I definitely need to rework my explanation to be more clear.  It looks like you were assuming that the note referenced real subroutines within a reportwriter section and not being able to mock them.  What I was specifically referring to was calling the _framework_ mocking APIs within a reportwriter.  Such an example might be:
```
; In the context of a unit test, I have a record structure, and I want to insert the record
; structure items into my mock table
select into "nl:"
from (dummyt d1 with seq = value(4))
plan d1
detail
    call cclutAddMockData("person", 
        concat(test_record->qual[d1.seq]->id, "|", test_record->qual[d1.seq]->name))
with nocounter
```

Second, you are correct in that it's not the _entire_ mocking API that won't work.  In my head, I was thinking "table mocking API", but I just wrote "mocking API".  The reason for that is that the implementation for the table mocking API uses DDL and DML which are not supported from reportwriter sections, but other stuff (like cclutDefineMockEntity or cclutRemoveMockEntity) would be fine.  I think as long as I add further clarification that it's referencing calling table mocking API functions, then I don't think it will be necessary to have specific examples (since it would just be a list of the table mocking functions).

Let me do some more work cleaning up this statement and see what you think after my next push.  We can always add examples if you still feel that it's unclear.

Reply:
I simply misinterpreted "The mocking API calls cannot be used...." as "Mocks cannot be used...". Fatigue could have been a factor, but  perhaps something like follows.

"It might be tempting to use a dummyt query to set up mock data from a record structure, but various mocking calls such as cclutCreateMockTable, cclutDropMockTable and cclutAddMockData cannot be executed within the context of a query (because the implementations execute queries). Use a for loop instead."

Reply:
Understood.  Will do.